problem-gson/src/test/java/org/zalando/problem/GsonProblemTest.java

```diff
@@ -0,0 +1,296 @@
+
+import java.io.*;
+import java.net.URI;
+import java.net.URL;
+import java.util.List;
+import java.util.Objects;
+
+import static com.jayway.jsonassert.JsonAssert.with;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.hasEntry;
+import static org.hamcrest.Matchers.hasSize;
+import static org.hamcrest.Matchers.hasToString;
+import static org.hamcrest.Matchers.instanceOf;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.notNullValue;
+import static org.hamcrest.Matchers.nullValue;
+import static org.hamcrest.Matchers.startsWith;
+import static org.hobsoft.hamcrest.compose.ComposeMatchers.hasFeature;
+import static org.zalando.problem.Status.BAD_REQUEST;
+
+class GsonProblemTest {
+
+    private static Gson gson;
+
+    @BeforeAll
+    static void init() {
+        GsonBuilder builder = new GsonBuilder();
+        // Register known sub-types
+        ProblemAdapterFactory problemAdapter = new ProblemAdapterFactory()
+                .registerSubtype(InsufficientFundsProblem.class, InsufficientFundsProblem.TYPE)
+                .registerSubtype(OutOfStockException.class, OutOfStockException.TYPE);
+        builder.registerTypeAdapterFactory(problemAdapter);
+        gson = builder.create();
+    }
+
+    private static JsonReader getReader(final String name) throws FileNotFoundException {
+        final ClassLoader loader = Thread.currentThread().getContextClassLoader();
+        URL resource = Objects.requireNonNull(loader.getResource(name), () -> "resource " + name + " not found.");
+        return new JsonReader(new FileReader(new File(resource.getPath())));
+    }
+
+    private String getStackTrace(final Throwable throwable) {
+        final StringWriter writer = new StringWriter();
+        throwable.printStackTrace(new PrintWriter(writer));
+        return writer.toString();
+    }
+
+    @Test
+    void shouldSerializeDefaultProblem() {
+        final Problem problem = Problem.valueOf(Status.NOT_FOUND);
+        final String json = gson.toJson(problem);
+
+        with(json)
+                .assertThat("$.*", hasSize(2))
+                .assertThat("$.title", is("Not Found"))
+                .assertThat("$.status", is(404));
+    }
+
+    @Test
+    void shouldSerializeCustomProperties() {
+        final Problem problem = Problem.builder()
+                .withType(URI.create("https://example.org/out-of-stock"))
+                .withTitle("Out of Stock")
+                .withStatus(BAD_REQUEST)
+                .withDetail("Item B00027Y5QG is no longer available")
+                .with("product", "B00027Y5QG")
+                .build();
+
+        final String json = gson.toJson(problem);
+
+        with(json)
+                .assertThat("$.*", hasSize(5))
+                .assertThat("$.product", is("B00027Y5QG"));
+    }
+
+    // FIXME Custom serialization not working
+    void shouldSerializeCustomProblem() {
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use explicit scoping instead of the default package private level](https://app.codacy.com/app/whiskeysierra/problem/pullRequest?prid=4476243)